app/src/main/java/com/github/epfl/meili/forum/FirebasePostService.kt

```diff
@@ -0,0 +1,75 @@
+package com.github.epfl.meili.forum
+
+import android.util.Log
+import com.github.epfl.meili.forum.Post.Companion.toPost
+import com.google.firebase.firestore.DocumentChange
+import com.google.firebase.firestore.FirebaseFirestore
+import java.util.*
+
+class FirebasePostService() : PostService() {
+
+    companion object {
+        val DEFAULT_DB = { FirebaseFirestore.getInstance() }
+
+        // Change this for dependency injection
+        var dbProvider: () -> FirebaseFirestore = DEFAULT_DB
+    }
+
+    private val TAG = "FirebasePostService"
+    private var db: FirebaseFirestore = dbProvider()
+
+    init {
+        // Listen to changes in post collection and make appropriate changes
+        db.collection("posts").addSnapshotListener { snapshot, e ->
+            // Handle errors
+            if (e != null || snapshot == null) {
+                Log.w(TAG, "Listen failed.", e)
+                return@addSnapshotListener
+            }
+
+            // Modify local list for each change in Firestore database
+            for (post in snapshot!!.documentChanges) {
+                when (post.type) {
+                    DocumentChange.Type.ADDED -> posts.add(post.document.toPost()!!)
+                    DocumentChange.Type.MODIFIED -> {
+                        posts.remove(post.document.toPost())
+                        posts.add(post.document.toPost()!!)
+                    }
+                    DocumentChange.Type.REMOVED -> posts.remove(post.document.toPost())
+                }
+            }
+
+            notifyObservers()
+        }
+    }
+
+    /** Get Post data from its id */
+    override fun getPostFromId(id: String?): Post? {
+        for (p in posts) {
+            if (p.id == id) {
+                return p
+            }
+        }
+
+        return null
+    }
```

Comment:
Just a minor detail but maybe if it prevails to be problematic we can make posts a hash table or a b-tree just to avoid that O(n) complexity.

Reply:
True, I will add in the backlog so we have it somewhere